File: java/05-reactor/src/test/java/com/fResult/reactor/ch5_05/HandleTest.java
package com.fResult.reactor.ch5_05;

import java.util.stream.Stream;
import lombok.extern.log4j.Log4j2;
import lombok.val;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Flux;
import reactor.test.StepVerifier;

@Log4j2
class HandleTest {
  @Test
  void handle() {
    StepVerifier.create(emitUntilErrorValue(5, 4))
        .expectNext(0, 1, 2, 3)
        .expectError(IllegalArgumentException.class)
        .verify();

    StepVerifier.create(emitUntilErrorValue(3, 3)).expectNext(0, 1, 2).verifyComplete();
  }

  private Flux<Integer> emitUntilErrorValue(int max, int errorAt) {
    return Flux.range(0, max)
        .handle(
            (value, sink) -> {
              val upTo = Stream.iterate(0, i -> i < errorAt, i -> i + 1).toList();

              if (upTo.contains(value)) {
                sink.next(value);
                return;
              }
              if (value == errorAt) {
                sink.error(new IllegalArgumentException("No " + errorAt + " for you!"));
                return;
              }

              sink.complete();
            });
  }
}

File: kotlin/05-reactor/src/test/kotlin/com/fResult/reactor/ch5_05/HandleTest.kt
package com.fResult.reactor.ch5_05

import org.apache.logging.log4j.LogManager
import reactor.core.publisher.Flux
import reactor.test.StepVerifier
import java.util.stream.Stream
import kotlin.test.Test

class HandleTest {
  private val log = LogManager.getLogger(HandleTest::class.java)

  @Test
  fun handle() {
    StepVerifier.create(emitUntilErrorValue(5, 4))
      .expectNext(0, 1, 2, 3)
      .expectError(IllegalArgumentException::class.java)
      .verify()

    StepVerifier.create(emitUntilErrorValue(3, 3)).expectNext(0, 1, 2).verifyComplete()
  }

  private fun emitUntilErrorValue(max: Int, errorAt: Int): Flux<Int> = Flux.range(0, max)
    .handle { value, sink ->
      val upTo = Stream.iterate(0, { it < errorAt }) { it + 1 }.toList()

      if (upTo.contains(value)) {
        sink.next(value)
        return@handle
      }
      if (value == errorAt) {
        sink.error(IllegalArgumentException("No 4 for you!"))
        return@handle
      }
      sink.complete()
    }
}
